walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,169 @@
+    Set<String> attendees = new HashSet<>(request.getAttendees());
+    long requestedDuration = request.getDuration();
+
+    if (Long.compare(requestedDuration, TimeRange.WHOLE_DAY.duration()) > 0) {
+      return Arrays.asList();
+    } else if (events.isEmpty() || attendees.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    Collection<TimeRange> occupiedTimeRanges =
+        getRelevantTimeRangesFromEvents(events, attendees);
+
+    return getAvailabeTimeRanges(occupiedTimeRanges, requestedDuration);
+  }
+
+  /**
+  * Finds the available {@code TimeRange} having at least a length of {@code requestedDuration},
+  * given some already occupied {@code TimeRange}. The resultant list of available
+  * {@code TimeRange} is sorted in ascending chronological order.
+  */
+  private List<TimeRange> getAvailabeTimeRanges(Collection<TimeRange> occupiedTimeRanges,
+      long requestedDuration) {
+    if (occupiedTimeRanges.isEmpty()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    List<TimeRange> availableTimeRanges = new LinkedList<>();
+    List<TimeRange> startOrderedTimeRanges = getOrderedTimeRanges(occupiedTimeRanges,
+                                                                  TimeRange.ORDER_BY_START);
+    List<TimeRange> endOrderedTimeRanges = getOrderedTimeRanges(occupiedTimeRanges,
+                                                                TimeRange.ORDER_BY_END);
+
+    // Potentially add the {@code TimeRange} before the first existing {@code TimeRange}.
+    TimeRange firstStartingTimeRange = startOrderedTimeRanges.get(0);
+    checkDurationAndAddAvailableTimeRange(requestedDuration,
+                                          TimeRange.START_OF_DAY, firstStartingTimeRange.start(),
+                                          availableTimeRanges);
+
+    // Find available {@code TimeRange} by looking at the available time between one ending
+    // {@code TimeRange} and the immediately following starting {@code TimeRange}.
+    ListIterator<TimeRange> startOrderedTimeRangesIterator = startOrderedTimeRanges.listIterator();
+    ListIterator<TimeRange> endOrderedTimeRangesIterator = endOrderedTimeRanges.listIterator();
+    TimeRange endingTimeRange = endOrderedTimeRangesIterator.next();
+    TimeRange startingTimeRange = startOrderedTimeRangesIterator.next();
+    while (true) {
+      // There is no available time after the current ending {@code TimeRange} and before the
+      // immediately following starting {@code TimeRange}.
+      if (endingTimeRange.overlaps(startingTimeRange)) {
+        if (startOrderedTimeRangesIterator.hasNext()) {
+          startingTimeRange = startOrderedTimeRangesIterator.next();
+        } else {
+          break;
+        }
+        continue;
+      }
+      // There is another {@code TimeRange} that ends later than the current ending
+      // {@code TimeRange} + before the immediately following, starting {@code TimeRange}.
```

Comment:
I see you mention "before the immediately following, starting" a few times, but I'm not sure what you mean? Do you mean before the next available starting Timerange?

Reply:
Yes

Reply:
I'll reword the comments. Thanks for pointing this out!